lib/ringbuf/src/stringbuf.rs

```diff
@@ -0,0 +1,159 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+pub use core::fmt;
+pub use core::fmt::Write;
+/// Re-export the bits we use so that code generated by the
+/// macros is guaranteed to be able to find them.
+pub use userlib::util::StaticCell;
+
+/// Declares a ringbuffer in the current module or context.
+///
+/// `stringbuf!(NAME, N, expr)` makes a ringbuffer named `NAME`,
+/// with room for `N` such entries, all of which are initialized to `expr`.
+///
+/// The resulting ringbuffer will be static, so `NAME` should be uppercase. If
+/// you want your ringbuffer to be detected by Humility's automatic scan, its
+/// name should end in `STRINGBUF`.
+///
+/// The actual type of `name` will be `StaticCell<Stringbuf<N>>`.
+///
+/// To support the common case of having one quickly-installed ringbuffer per
+/// module, if you omit the name, it will default to `LOG__STRINGBUF`.
+#[cfg(not(feature = "disabled"))]
+#[macro_export]
+macro_rules! stringbuf {
+    ($name:ident, $n:expr, $init:expr) => {
+        #[used]
+        static $name: $crate::StaticCell<$crate::stringbuf::Stringbuf<$n>> =
+            $crate::StaticCell::new($crate::stringbuf::Stringbuf {
+                last: None,
+                buffer: [0; $n],
+            });
+    };
+    ($n:expr, $init:expr) => {
+        $crate::stringbuf!(LOG__STRINGBUF, $n, $init);
+    };
+}
+
+#[cfg(not(feature = "disabled"))]
+#[macro_export]
+macro_rules! stringbuf_root {
```

Comment:
So, what is the difference between `stringbuf_root` and `stringbuf`? I see that `stringbuf_root` is marked pub whereas `stringbuf` isn't. Is that the use case i.e. use `stringbuf_root` if the ring buffer is supposed to be shared outside of a module? The suffix "_root" doesn't really make it obvious though.

Reply:
So having both `stringbuf` and `stringbuf_root` macros was a leftover artifact.  I cleaned that up so there is just one `stringbuf` macro that initializes. See this[ new PR](https://github.com/rivosinc/hubris/pull/10).
The root postfix refers to the `stringbuf` at the root of the `stringbuf` module.  This is just a naming convention borrowed from `ringbuf`.

Reply:
ack